src/EditorFeatures/Core/Implementation/IntelliSense/AsyncCompletion/ItemManager.CompletionListUpdater.cs

```diff
@@ -0,0 +1,846 @@
+using System.Diagnostics;
+using System.Globalization;
+using System.Linq;
+using System.Threading;
+using Microsoft.CodeAnalysis.Completion;
+using Microsoft.CodeAnalysis.Options;
+using Microsoft.CodeAnalysis.PatternMatching;
+using Microsoft.CodeAnalysis.PooledObjects;
+using Microsoft.CodeAnalysis.Shared.Extensions;
+using Microsoft.CodeAnalysis.Text;
+using Microsoft.CodeAnalysis.Text.Shared.Extensions;
+using Microsoft.VisualStudio.Language.Intellisense.AsyncCompletion;
+using Microsoft.VisualStudio.Language.Intellisense.AsyncCompletion.Data;
+using Microsoft.VisualStudio.Text;
+using Roslyn.Utilities;
+using RoslynCompletionItem = Microsoft.CodeAnalysis.Completion.CompletionItem;
+using VSCompletionItem = Microsoft.VisualStudio.Language.Intellisense.AsyncCompletion.Data.CompletionItem;
+
+namespace Microsoft.CodeAnalysis.Editor.Implementation.IntelliSense.AsyncCompletion
+{
+    internal partial class ItemManager
+    {
+        /// <summary>
+        /// Handles the filtering, sorting and selection of the completion items based on user inputs 
+        /// (e.g. typed characters, selected filters, etc.)
+        /// </summary>
+        private sealed class CompletionListUpdater
+        {
+            private readonly IAsyncCompletionSession _session;
+            private readonly AsyncCompletionSessionDataSnapshot _data;
+            private readonly RecentItemsManager _recentItemsManager;
+            private readonly IGlobalOptionService _globalOptions;
+            private readonly CancellationToken _cancellationToken;
+
+            private readonly bool _hasSuggestedItemOptions;
+            private readonly string _filterText;
+            private readonly CompletionTriggerKind _initialRoslynTriggerKind;
+            private readonly Document? _document;
+            private readonly CompletionService? _completionService;
+            private readonly CompletionRules _completionRules;
+            private readonly CompletionHelper _completionHelper;
+            private readonly bool _highlightMatchingPortions;
+
+            private CompletionTriggerReason TriggerReason => _data.Trigger.Reason;
+            private CompletionFilterReason FilterReason => Helpers.GetFilterReason(_data.Trigger);
+
+            private Func<ImmutableArray<(RoslynCompletionItem, PatternMatch?)>, string, ImmutableArray<RoslynCompletionItem>> FilterMethod
+                => _completionService == null
+                    ? ((itemsWithPatternMatches, text) => CompletionService.FilterItems(_completionHelper, itemsWithPatternMatches, text))
+                    : ((itemsWithPatternMatches, text) => _completionService.FilterItems(_document!, itemsWithPatternMatches, text));
+
+            // PERF: Create a singleton to avoid lambda allocation on hot path
+            private static readonly Func<TextSpan, RoslynCompletionItem, Span> s_highlightSpanGetter
+                = (span, item) => span.MoveTo(item.DisplayTextPrefix?.Length ?? 0).ToSpan();
```

Comment:
feels like we can just move this to where it it used (and keep it static to avoid the allocation).

Reply:
still applies?

Reply:
Missed this one. Fixed